plugins/org.python.pydev/src/org/python/pydev/editor/correctionassist/heuristics/AssistFString.java

```diff
@@ -0,0 +1,181 @@
+import org.python.pydev.shared_core.string.FastStringBuffer;
+import org.python.pydev.shared_core.string.StringUtils;
+
+public class AssistFString implements IAssistProps {
+    @Override
+    public List<ICompletionProposalHandle> getProps(PySelection ps, IImageCache imageCache, File f,
+            IPythonNature nature, IPyEdit edit, int offset) throws BadLocationException, MisconfigurationException {
+        PySourceViewer viewer = null;
+        if (edit instanceof PyEdit) { //only in tests it is actually null
+            viewer = ((PyEdit) edit).getPySourceViewer();
+        }
+        List<ICompletionProposalHandle> lst = new ArrayList<ICompletionProposalHandle>();
+
+        IDocument doc = ps.getDoc();
+        ITypedRegion partition = ((FastPartitioner) PyPartitionScanner.checkPartitionScanner(doc)).getPartition(offset);
+        if (!ParsingUtils.isStringContentType(partition.getType())) {
+            return lst;
+        }
+
+        final String stringPartitionContents = doc.get(partition.getOffset(), partition.getLength());
+        int formatCount = StringUtils.count(stringPartitionContents, "%s");
+        if (formatCount == 0) {
+            return lst;
+        }
+
+        // get string end offset
+        int partitionEndOffset = partition.getOffset() + partition.getLength();
+        // get format end offset
+        int i = partitionEndOffset;
+        try {
+            ParsingUtils parsingUtils = ParsingUtils.create(doc);
+            while (i < parsingUtils.len()) {
+                char c = parsingUtils.charAt(i);
+                if (Character.isWhitespace(c)) {
+                    i++;
+                    continue;
+                }
+
+                if (c == '{') {
+                    return lst;
+                }
+
+                if (c == '[' || c == '(' || c == '"' || c == '\'') {
+                    i = parsingUtils.eatPar(i, null, c) + 1;
+                    break;
+                }
+                if (Character.isJavaIdentifierPart(c)) {
+                    i++;
+                    while (i < parsingUtils.len()) {
+                        c = parsingUtils.charAt(i);
+                        if (Character.isJavaIdentifierPart(c) || c == '.') {
+                            i++;
+                            continue;
+                        } else if (c == '{' || c == '[' || c == '(' || c == '"' || c == '\'') {
+                            i = parsingUtils.eatPar(i, null, c) + 1;
+                            c = parsingUtils.charAt(i);
+                            while (Character.isWhitespace(c)) {
+                                i++;
+                                c = parsingUtils.charAt(i);
+                            }
+                            if (c == '.') {
+                                i++;
+                                continue;
+                            }
+                            break; // inner break
+                        }
+                    }
```

Comment:
I think you have a case where `i` is not incremented here which can lead to an infinite loop.

I believe you can reproduce it with something as:

`'%s %s' % foo, bar`.

I also realized we aren't supporting this case (which is pretty common), so, it's something to be added to the tests too.

We should also support things as:

`'%s %s' % a().foo, bar[1]`.

Reply:
As a note, we should collect whitespaces and commas in this case but should break on new lines (\r or \n).

Reply:
I ran tests here and I thought Python did not support 2 or more arguments without enclosing parenthesis to format a string...

Reply:
It does if you do something as: `call("%s - %s" % foo, bar)`

Reply:
The parser accepts it if it's inside the `(...)`